cpp/src/io/parquet/page_enc.cu

```diff
@@ -0,0 +1,1550 @@
+    }
+}
+
+
+/**
+ * Minimal thrift compact protocol support
+ **/
+inline __device__ uint8_t *cpw_put_uint32(uint8_t *p, uint32_t v) {
+    while (v > 0x7f) {
+        *p++ = v | 0x80;
+        v >>= 7;
+    }
+    *p++ = v;
+    return p;
+}
+
+inline __device__ uint8_t *cpw_put_uint64(uint8_t *p, uint64_t v) {
+    while (v > 0x7f) {
+        *p++ = v | 0x80;
+        v >>= 7;
+    }
+    *p++ = v;
+    return p;
+}
+
+inline __device__ uint8_t *cpw_put_int32(uint8_t *p, int32_t v) {
+    int32_t s = (v < 0);
+    return cpw_put_uint32(p, (v ^ -s) * 2 + s);
+}
+
+inline __device__ uint8_t *cpw_put_int64(uint8_t *p, int64_t v) {
+    int64_t s = (v < 0);
+    return cpw_put_uint64(p, (v ^ -s) * 2 + s);
+}
+
+inline __device__ uint8_t *cpw_put_fldh(uint8_t *p, int f, int cur, int t) {
+    if (f > cur && f <= cur + 15) {
+        *p++ = ((f - cur) << 4) | t;
+        return p;
+    }
+    else {
+        *p++ = t;
+        return cpw_put_int32(p, f);
+    }
+}
+
+#define CPW_BEGIN_STRUCT(hdr_start) {               \
+    uint8_t *p = hdr_start;                         \
+    int cur_fld = 0;                                \
+
+#define CPW_FLD_STRUCT_BEGIN(f)                     \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_STRUCT); \
+    cur_fld = 0;                                    \
+
+#define CPW_FLD_STRUCT_END(f)                       \
+    *p++ = 0;                                       \
+    cur_fld = f;                                    \
+
+#define CPW_FLD_INT32(f, v)                         \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_I32);    \
+    p = cpw_put_int32(p, v);                        \
+    cur_fld = f;                                    \
+
+#define CPW_FLD_INT64(f, v)                         \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_I64);    \
+    p = cpw_put_int64(p, v);                        \
+    cur_fld = f;                                    \
+
+#define CPW_FLD_BINARY(f, v, l)                     \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_BINARY); \
+    p = cpw_put_uint32(p, l);                       \
+    memcpy(p, v, l);                                \
+    p += l;                                         \
+    cur_fld = f;                                    \
+
+#define CPW_END_STRUCT(hdr_end)                     \
+    *p++ = 0;                                       \
+    hdr_end = p;                                    \
+    }                                               \
+
+#define CPW_END_STRUCT_NOTERMINATION(hdr_end)       \
+    hdr_end = p;                                    \
+    }                                               \
+
+
+__device__ uint8_t *EncodeStatistics(uint8_t *start, const statistics_chunk *s, const EncColumnDesc *col, float *fp_scratch)
+{
+    uint8_t *end, dtype, dtype_len;
+    dtype = col->stats_dtype;
+    switch (dtype) {
+    case dtype_bool8:
+    case dtype_int8:
+        dtype_len = 1;
```

Comment:
The `ArrayOutOfBoundsException` error from the Java parquet loaders seems to be triggered by this code.  Those loaders expect the statistics values to at least be 4-byte values.  Makes sense to me, as I don't believe Parquet has support for INT8 or INT16 data types directly.

Using `dtype_len = 4` for the int8 and int16 cases fixes the footer parsing problem in parquet-mr tools and Spark and allows both to decode the column values properly.

Reply:
Awesome, will make the change.

Reply:
Makes sense, as it is based on the physical type. I'm assuming bool remains 1-byte, though ?

Reply:
Yes, boolean statistics are expecting a 1-byte value.